monitor/guard.go

```diff
@@ -0,0 +1,152 @@
+		event := NewEventFromBytes(iterator.Value())
+		log.Infoln("Process pusher event", event.Name)
+		err = m.db.Delete(iterator.Key())
+		if err != nil {
+			log.Errorln("db Delete err", err)
+			continue
+		}
+
+		switch e := event.ParseEvent(m.ethClient).(type) {
+		case *mainchain.CelerLedgerIntendSettle:
+			m.guardIntendSettle(e)
+		case *mainchain.CelerLedgerIntendWithdraw:
+			m.guardIntendWithdrawChannel(e)
+		}
+	}
+}
+
+func (m *Monitor) guardIntendSettle(intendSettle *mainchain.CelerLedgerIntendSettle) {
+	log.Infof("Guard IntendSettle %x, tx hash %x", intendSettle.ChannelId, intendSettle.Raw.TxHash)
+	requests := m.getRequests(intendSettle.ChannelId)
+	for _, request := range requests {
+		m.guardRequest(request, intendSettle.Raw, IntendSettle)
+	}
+}
+
+func (m *Monitor) guardIntendWithdrawChannel(intendWithdrawChannel *mainchain.CelerLedgerIntendWithdraw) {
+	log.Infof("Guard intendWithdrawChannel %x, tx hash %x", intendWithdrawChannel.ChannelId, intendWithdrawChannel.Raw.TxHash)
+	requests := m.getRequests(intendWithdrawChannel.ChannelId)
+	for _, request := range requests {
+		m.guardRequest(request, intendWithdrawChannel.Raw, IntendWithdrawChannel)
+	}
+}
+
+func (m *Monitor) getRequests(cid [32]byte) (requests []subscribe.Request) {
+	channelId := cid[:]
+	addresses, seqNums, err := m.ethClient.Ledger.GetStateSeqNumMap(
+		&bind.CallOpts{BlockNumber: sdk.NewIntFromUint64(m.secureBlkNum).BigInt()},
+		cid)
+	if err != nil {
+		log.Errorln("Query StateSeqNumMap err", err)
+		return
+	}
+
+	for _, addr := range addresses {
+		peerFrom := mainchain.Addr2Hex(addr)
+		request, err := m.getRequest(channelId, peerFrom)
+		if err != nil {
+			continue
+		}
+
+		if seqNums[request.PeerFromIndex].Uint64() >= request.SeqNum {
+			log.Infoln("Ignore the intendSettle event with an equal or larger seqNum")
+			continue
+		}
+
+		requests = append(requests, request)
+	}
+
+	return requests
+}
+
+func (m *Monitor) guardRequest(request subscribe.Request, rawLog ethtypes.Log, eventName EventName) {
+	log.Infoln("Guard request", request)
+	if request.GuardTxHash != "" {
+		log.Errorln("Request has been fulfilled")
+		return
+	}
+
+	if !m.isRequestGuard(request, rawLog.BlockNumber) {
+		log.Infof("Not valid guard at current mainchain block")
+		event := NewEvent(IntendSettle, rawLog)
+		err := m.db.Set(GetGuardKey(rawLog), event.MustMarshal())
+		if err != nil {
+			log.Errorln("db Set err", err)
+		}
+		return
+	}
+
+	var signedSimplexState chain.SignedSimplexState
+	err := proto.Unmarshal(request.SignedSimplexStateBytes, &signedSimplexState)
+	if err != nil {
+		log.Errorln("Unmarshal SignedSimplexState error:", err)
+		return
+	}
+
+	signedSimplexStateArrayBytes, err := proto.Marshal(&chain.SignedSimplexStateArray{
+		SignedSimplexStates: []*chain.SignedSimplexState{&signedSimplexState},
+	})
+	if err != nil {
+		log.Errorln("Marshal signedSimplexStateArrayBytes error:", err)
+		return
+	}
+
+	// TODO: should not waitmined
```

Comment:
is TODO still valid?

Reply:
Yes, we should use the callback method to allow concurrent pending transactions, and handle the failure more properly, e.g., retry on tx dropped